dislib/model_selection/_split.py

```diff
@@ -0,0 +1,229 @@
+        If RandomState instance, random_state is the random number generator;
+        If None, the random number generator is the RandomState instance used
+        by `np.random`. Used when ``shuffle`` == True.
+    """
+
+    def __init__(self, n_splits=5, shuffle=True, random_state=None):
+        if not isinstance(n_splits, numbers.Integral):
+            raise ValueError('The number of folds must be of Integral type. '
+                             '%s of type %s was passed.'
+                             % (n_splits, type(n_splits)))
+        n_splits = int(n_splits)
+
+        if n_splits <= 1:
+            raise ValueError(
+                "k-fold cross-validation requires at least one"
+                " train/test split by setting n_splits=2 or more,"
+                " got n_splits={0}.".format(n_splits))
+
+        if not isinstance(shuffle, bool):
+            raise TypeError("shuffle must be True or False;"
+                            " got {0}".format(shuffle))
+
+        self.n_splits = n_splits
+        self.shuffle = shuffle
+        self.random_state = random_state
+
+    def split(self, x, y=None):
+        """Generates K-fold splits.
+
+        Parameters
+        ----------
+        x : ds_array
+            Samples array.
+        y : ds_array, optional (default=None)
+            Corresponding labels or values.
+
+        Yields
+        ------
+        train_data : train_x, train_y
+            The training ds_arrays for that split. If y is None, train_y is
+            None.
+        test_data : test_x, test_y
+            The testing ds_arrays data for that split. If y is None, test_y is
+            None.
+        """
+        k = self.n_splits
+        if self.shuffle:
+            shuffled = utils.shuffle(x, y, self.random_state)
+            if y is None:
+                x = shuffled
+            else:
+                x, y = shuffled
+        n_total = x.shape[0]
+        n_each_section, extras = divmod(n_total, k)
+        section_sizes = np.empty((k,), dtype=int)
+        section_sizes[:extras] = n_each_section + 1
+        section_sizes[extras:] = n_each_section
+        div_points = np.cumsum(section_sizes)
+        yield get_kfold_partition(x, y, 0, div_points[0])
+        for i in range(1, k):
+            yield get_kfold_partition(x, y, div_points[i - 1], div_points[i])
+
+    def get_n_splits(self):
+        """Get the number of CV splits that this splitter does.
+
+        Returns
+        ------
+        n_splits : int
+            The number of splits performed by this CV splitter.
+        """
+        return self.n_splits
+
+
+def get_kfold_partition(x, y, start, end):
+    train_x = merge_slices(x[:start], x[end:])
+    test_x = x[start:end]
+    train_y = None
+    test_y = None
+    if y is not None:
+        train_y = merge_slices(y[:start], y[end:])
+        test_y = y[start:end]
+    return (train_x, train_y), (test_x, test_y)
+
+
+def merge_slices(s1, s2):
```

Comment:
Maybe add some comments to better understand what is happening inside this function.

Reply:
done!